runtime/remote_config.go

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zanzibar
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/buger/jsonparser"
+	"github.com/pkg/errors"
+	"github.com/uber-go/tally"
+	"go.uber.org/zap"
+)
+
+// RemoteConfig allow accessing values from a periodically refreshed json file
+type RemoteConfig interface {
+	GetBoolean(string) (bool, error)
+	GetFloat(string, float64) (float64, error)
+	GetInt(string, int64) (int64, error)
+	GetString(string, string) (string, error)
+	GetStruct(string, interface{}) error
+	Subscribe(string, *func())
```

Comment:
in `Subscribe()` we need to know which keys changed and which did not.

Reply:
mm ... I maybe misunderstood.

current implementation assumes users `know` exactly what key they're listening and they just `remoteConfig.get("that-key")` whenever config reloads, something like:

```
remoteConfig.Subscribe("chao", &func(){
       v, err := remoteConfig.get("some-key")
      // handle err
       reloadNewConfig(v)
})
```

am I getting you right that we actually want better granularity when key changes? like:

```
remoteConfig.Subscribe("chao", "some-key", &func(){
       v, err := remoteConfig.get("some-key")
      // handle err
       reloadNewConfig(v)
})
```

Reply:
we need to know what has changed so that if our key has NOT changed we do not re-read it for no reason.